File: modules/10-image-generation/py/param_sweep.py
"""param_sweep.py — Task 2 🟡: explore guidance scale, steps, and negative prompts.

What it teaches:
    Every "knob" on an image generation request has a concrete effect rooted
    in the math. This script sends ~6 API calls across two parameter axes and
    assembles the results into a PNG grid so you can see the differences
    side-by-side.

    Parameters explored:
        guidance_scale  — how strongly the denoiser follows the text prompt
                          (classifier-free guidance weight, see README Concept 5)
        num_inference_steps — how many denoising steps to run
                          (more ≈ more detail, diminishing returns past ~40)

    The negative_prompt is kept constant and its effect explained below.

How to run (from the repo root):
    uv run python modules/10-image-generation/py/param_sweep.py

    Sends ~6 API requests. Each takes a few seconds. Total ~30-60 s.

Requires: REPLICATE_API_TOKEN (or IMAGE_PROVIDER + its key)
Requires: pip install pillow  (for grid assembly; part of --extra imagegen)

Note on negative prompts:
    Classifier-free guidance (CFG) runs the U-Net TWICE per step:
        ε_guided = ε_uncond + scale * (ε_cond - ε_uncond)
    The negative prompt is encoded by CLIP and used as ε_uncond, making the
    model steer AWAY from those features. An empty negative prompt uses a zero
    embedding instead. Common values: "blurry, low quality, watermark, deformed".
"""

from __future__ import annotations

import io
import sys
from itertools import product

from image_client import GenerateOptions, generate_image

# ---------------------------------------------------------------------------
# Sweep configuration — change these to explore
# ---------------------------------------------------------------------------

PROMPT = (
    "A majestic snow leopard on a mountain ridge at dusk, cinematic lighting, "
    "ultra-detailed fur, professional wildlife photography"
)

# Try: ""  vs  "blurry, low quality, watermark, cartoon, distorted"
NEGATIVE_PROMPT = "blurry, low quality, watermark, cartoon, deformed"

# Axis 1: guidance_scale (columns)
GUIDANCE_SCALES = [3.0, 7.5, 15.0]

# Axis 2: num_inference_steps (rows)
STEPS_LIST = [10, 30]

# Fixed seed so all images share the same noise starting point —
# differences come only from the parameter being swept.
SEED = 42

OUTPUT_PATH = "sweep_grid.png"


def _assemble_grid(images: list[bytes], cols: int, rows: int) -> bytes:
    """Stitch PIL images into a (cols × rows) grid. Returns PNG bytes."""
    try:
        from PIL import Image, ImageDraw, ImageFont
    except ImportError:
        print(
            "\nPillow is not installed. Run `uv sync --extra imagegen` to get it,\n"
            "or install manually: pip install pillow\n"
            "Saving individual images instead."
        )
        for idx, img_bytes in enumerate(images):
            path = f"sweep_img_{idx}.png"
            with open(path, "wb") as f:
                f.write(img_bytes)
            print(f"  Saved {path}")
        sys.exit(0)

    pil_images = [Image.open(io.BytesIO(b)) for b in images]
    w, h = pil_images[0].size
    label_h = 30  # px for axis labels

    grid_w = cols * w
    grid_h = rows * h + label_h
    grid = Image.new("RGB", (grid_w, grid_h), (240, 240, 240))

    # Try to get a font; fall back gracefully
    try:
        font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 18)
    except Exception:
        font = ImageFont.load_default()

    draw = ImageDraw.Draw(grid)

    for idx, pil_img in enumerate(pil_images):
        col = idx % cols
        row = idx // cols
        x = col * w
        y = row * h + label_h
        grid.paste(pil_img, (x, y))

    # Column labels (guidance scale)
    for col, gs in enumerate(GUIDANCE_SCALES):
        draw.text((col * w + 5, 5), f"CFG={gs}", fill=(30, 30, 30), font=font)

    buf = io.BytesIO()
    grid.save(buf, format="PNG")
    return buf.getvalue()


def main() -> None:
    print(f'Prompt: "{PROMPT}"')
    print(f"Negative: {NEGATIVE_PROMPT!r}")
    print(f"Grid: {len(GUIDANCE_SCALES)} guidance scales × {len(STEPS_LIST)} step counts\n")

    combinations = list(product(STEPS_LIST, GUIDANCE_SCALES))
    images: list[bytes] = []

    for steps, gs in combinations:
        label = f"steps={steps}, cfg={gs}"
        print(f"  Generating {label} …")
        result = generate_image(
            GenerateOptions(
                prompt=PROMPT,
                negative_prompt=NEGATIVE_PROMPT,
                width=512,  # smaller for speed in a sweep
                height=512,
                guidance_scale=gs,
                num_inference_steps=steps,
                seed=SEED,
            )
        )
        images.append(result.image_bytes)
        print(f"  Done ({result.provider})")

    print(f"\nAssembling {len(GUIDANCE_SCALES)}×{len(STEPS_LIST)} grid …")
    grid_bytes = _assemble_grid(
        images, cols=len(GUIDANCE_SCALES), rows=len(STEPS_LIST)
    )
    with open(OUTPUT_PATH, "wb") as f:
        f.write(grid_bytes)
    print(f"Grid saved → {OUTPUT_PATH}")
    print()
    print("What to look for:")
    print("  CFG  3.0 (left col) : loosely follows the prompt; painterly, creative.")
    print("  CFG  7.5 (mid col)  : balanced — the default for most use cases.")
    print("  CFG 15.0 (right col): strong prompt adherence; may oversaturate/over-sharpen.")
    print("  10 steps (top row)  : faster but coarser; some coherence issues visible.")
    print("  30 steps (bot row)  : more detail and cleaner edges; diminishing returns above ~40.")


if __name__ == "__main__":
    main()

File: modules/10-image-generation/ts/param_sweep.ts
/**
 * param_sweep.ts — Task 2 🟡: explore guidance scale, steps, and negative prompts.
 *
 * What it teaches:
 *   Every "knob" on an image request has a concrete effect rooted in the math.
 *   This script sweeps two axes — guidance_scale (columns) and
 *   num_inference_steps (rows) — at a fixed seed, so the only thing that varies
 *   across a cell is the parameter being studied. It records EVERY parameter
 *   plus the seed for every cell in a reproducible sidecar JSON, so any run can
 *   be reconstructed exactly.
 *
 *   Parameters explored:
 *     guidanceScale       — how strongly the denoiser follows the text prompt
 *                           (classifier-free guidance weight, README Concept 5)
 *     numInferenceSteps   — how many denoising steps to run
 *                           (more ≈ more detail, diminishing returns past ~40)
 *   The negativePrompt is held constant; its effect is explained below.
 *
 * How to run (from the repo root):
 *   Hosted (sends ~6 API calls, needs a key):
 *     pnpm tsx modules/10-image-generation/ts/param_sweep.ts
 *   Deterministic offline stub (no key, no network — used by CI/tests):
 *     IMAGE_STUB=1 pnpm tsx modules/10-image-generation/ts/param_sweep.ts
 *
 * Note on negative prompts:
 *   Classifier-free guidance runs the U-Net TWICE per step:
 *     ε_guided = ε_uncond + scale * (ε_cond - ε_uncond)
 *   The negative prompt is encoded by CLIP and used as ε_uncond, steering the
 *   model AWAY from those features. Common value:
 *   "blurry, low quality, watermark, deformed".
 */

import { writeFileSync } from "node:fs";

import { generateImage, saveImage, type GenerateOptions } from "./image_client.ts";
import {
  encodePng,
  hashBytes,
  isOffline,
  solidGridPng,
  solidPixels,
  tintFor,
  type RGB,
} from "./stub_image.ts";

// ---------------------------------------------------------------------------
// Sweep configuration — change these to explore
// ---------------------------------------------------------------------------

export interface SweepConfig {
  prompt: string;
  negativePrompt: string;
  /** guidance_scale values — one column each. */
  guidanceScales: number[];
  /** num_inference_steps values — one row each. */
  stepsList: number[];
  /** Fixed seed so every cell shares the same noise starting point. */
  seed: number;
  width: number;
  height: number;
}

export const CONFIG: SweepConfig = {
  prompt:
    "A majestic snow leopard on a mountain ridge at dusk, cinematic lighting, " +
    "ultra-detailed fur, professional wildlife photography",
  // Try: ""  vs  "blurry, low quality, watermark, cartoon, deformed"
  negativePrompt: "blurry, low quality, watermark, cartoon, deformed",
  guidanceScales: [3.0, 7.5, 15.0], // columns
  stepsList: [10, 30], // rows
  seed: 42,
  width: 512, // smaller for speed in a sweep
  height: 512,
};

const GRID_OUTPUT = "sweep_grid.png";
const METADATA_OUTPUT = "sweep_metadata.json";

// ---------------------------------------------------------------------------
// Pure sweep model (deterministic — this is what the tests drive)
// ---------------------------------------------------------------------------

export interface SweepCell {
  row: number; // index into stepsList
  col: number; // index into guidanceScales
  steps: number;
  guidanceScale: number;
  seed: number;
  prompt: string;
  negativePrompt: string;
  width: number;
  height: number;
}

/**
 * Build the row-major grid of cells. Rows are step counts, columns are
 * guidance scales — the same ordering the Python `product(STEPS, GUIDANCE)`
 * uses. Every cell carries the full parameter set plus the shared seed.
 */
export function buildSweepCells(config: SweepConfig): SweepCell[] {
  const cells: SweepCell[] = [];
  config.stepsList.forEach((steps, row) => {
    config.guidanceScales.forEach((guidanceScale, col) => {
      cells.push({
        row,
        col,
        steps,
        guidanceScale,
        seed: config.seed,
        prompt: config.prompt,
        negativePrompt: config.negativePrompt,
        width: config.width,
        height: config.height,
      });
    });
  });
  return cells;
}

/** Map a cell to the provider-agnostic generate options. */
export function cellToOptions(cell: SweepCell): GenerateOptions {
  return {
    prompt: cell.prompt,
    negativePrompt: cell.negativePrompt,
    width: cell.width,
    height: cell.height,
    guidanceScale: cell.guidanceScale,
    numInferenceSteps: cell.steps,
    seed: cell.seed,
  };
}

/** Canonical identity string for a cell — the fingerprint the stub renders. */
export function cellKey(cell: SweepCell): string {
  return [
    cell.prompt,
    cell.negativePrompt,
    `${cell.width}x${cell.height}`,
    `cfg=${cell.guidanceScale}`,
    `steps=${cell.steps}`,
    `seed=${cell.seed}`,
  ].join("|");
}

/** Deterministic offline swatch colour for a cell. */
export function cellTint(cell: SweepCell): RGB {
  return tintFor(cellKey(cell));
}

/** Output filename for a cell's individual image. */
export function cellFilename(cell: SweepCell): string {
  return `sweep_cell_r${cell.row}_c${cell.col}.png`;
}

// ---------------------------------------------------------------------------
// Reproducible metadata
// ---------------------------------------------------------------------------

export interface SweepCellRecord {
  row: number;
  col: number;
  steps: number;
  guidanceScale: number;
  seed: number;
  output: string;
  provider: string;
  model: string;
  digest: string;
  tint: RGB | null;
}

export interface SweepMetadata {
  prompt: string;
  negativePrompt: string;
  seed: number;
  guidanceScales: number[];
  stepsList: number[];
  cols: number;
  rows: number;
  offline: boolean;
  grid: string | null;
  cells: SweepCellRecord[];
}

/**
 * Run the whole sweep deterministically offline. Returns the reproducible
 * metadata, the per-cell PNG bytes, and a composited grid PNG. No filesystem
 * or network access — the tests assert directly on the return value.
 */
export function buildSweepStub(config: SweepConfig): {
  metadata: SweepMetadata;
  cellImages: Uint8Array[];
  gridImage: Uint8Array;
} {
  const cells = buildSweepCells(config);
  const cellImages: Uint8Array[] = [];
  const tints: RGB[] = [];
  const records: SweepCellRecord[] = [];

  for (const cell of cells) {
    const tint = cellTint(cell);
    const pixels = solidPixels(cell.width, cell.height, tint);
    tints.push(tint);
    cellImages.push(encodePng(cell.width, cell.height, pixels));
    records.push({
      row: cell.row,
      col: cell.col,
      steps: cell.steps,
      guidanceScale: cell.guidanceScale,
      seed: cell.seed,
      output: cellFilename(cell),
      provider: "stub",
      model: "stub-solid-v1",
      digest: hashBytes(pixels),
      tint,
    });
  }

  const gridImage = solidGridPng(
    tints,
    config.guidanceScales.length,
    config.stepsList.length,
    config.width,
    config.height,
  );

  const metadata: SweepMetadata = {
    prompt: config.prompt,
    negativePrompt: config.negativePrompt,
    seed: config.seed,
    guidanceScales: config.guidanceScales,
    stepsList: config.stepsList,
    cols: config.guidanceScales.length,
    rows: config.stepsList.length,
    offline: true,
    grid: GRID_OUTPUT,
    cells: records,
  };

  return { metadata, cellImages, gridImage };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function runOffline(config: SweepConfig): Promise<void> {
  console.log("Offline stub mode (IMAGE_STUB/OFFLINE_SMOKE) — no API calls.\n");
  const { metadata, cellImages, gridImage } = buildSweepStub(config);

  metadata.cells.forEach((record, idx) => {
    saveImage(cellImages[idx], record.output);
  });
  saveImage(gridImage, GRID_OUTPUT);
  writeFileSync(METADATA_OUTPUT, JSON.stringify(metadata, null, 2));
  console.log(`Reproducible metadata → ${METADATA_OUTPUT}`);
}

async function runHosted(config: SweepConfig): Promise<void> {
  const cells = buildSweepCells(config);
  const records: SweepCellRecord[] = [];

  for (const cell of cells) {
    console.log(`  Generating steps=${cell.steps}, cfg=${cell.guidanceScale} …`);
    const result = await generateImage(cellToOptions(cell));
    const output = cellFilename(cell);
    saveImage(result.imageBytes, output);
    records.push({
      row: cell.row,
      col: cell.col,
      steps: cell.steps,
      guidanceScale: cell.guidanceScale,
      seed: cell.seed,
      output,
      provider: result.provider,
      model: result.model,
      digest: hashBytes(result.imageBytes),
      tint: null,
    });
  }

  const metadata: SweepMetadata = {
    prompt: config.prompt,
    negativePrompt: config.negativePrompt,
    seed: config.seed,
    guidanceScales: config.guidanceScales,
    stepsList: config.stepsList,
    cols: config.guidanceScales.length,
    rows: config.stepsList.length,
    offline: false,
    grid: null, // compositing hosted PNGs needs an image library; view cells individually
    cells: records,
  };
  writeFileSync(METADATA_OUTPUT, JSON.stringify(metadata, null, 2));
  console.log(`Reproducible metadata → ${METADATA_OUTPUT}`);
  console.log(
    "(Install an image library to composite the individual cells into a grid.)",
  );
}

async function main(): Promise<void> {
  console.log(`Prompt: "${CONFIG.prompt}"`);
  console.log(`Negative: ${JSON.stringify(CONFIG.negativePrompt)}`);
  console.log(
    `Grid: ${CONFIG.guidanceScales.length} guidance scales × ` +
      `${CONFIG.stepsList.length} step counts\n`,
  );

  if (isOffline()) {
    await runOffline(CONFIG);
  } else {
    await runHosted(CONFIG);
  }

  console.log("\nWhat to look for:");
  console.log(
    "  CFG  3.0 (left col) : loosely follows the prompt; painterly, creative.",
  );
  console.log("  CFG  7.5 (mid col)  : balanced — the default for most use cases.");
  console.log(
    "  CFG 15.0 (right col): strong adherence; may oversaturate/over-sharpen.",
  );
  console.log("  low steps (top row) : faster but coarser; some coherence issues.");
  console.log("  high steps (bot row): more detail; diminishing returns above ~40.");
}

// Run only when invoked directly (so the test file can import without executing).
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
